File: frontend/src/lib/keys.ts
/**
 * Keys module — the 24-key × 4-notation conversion TABLE from
 * planning/03-ui-design.md §6.6 (a table, NOT a formula — per spec), plus the
 * display formatter (Key Display As, §3.5) and the color lookup
 * (Colorful Keys, §3.5 + the §9 contrast-adjustment rule).
 *
 * Internal representation is flats-canonical everywhere (`Gbm`, 24 values);
 * the other three notations are pure display renames of the same key.
 */

import { KEY_COLORS } from "./palette";
import { ensureContrast } from "./color";

export type KeyNotation = "flats" | "sharps" | "camelot" | "openkey";

export interface KeyRow {
  flats: string;
  sharps: string;
  camelot: string;
  openkey: string;
}

/** §6.6 canonical conversion table, verbatim. Do not derive at runtime. */
export const KEY_TABLE = [
  { flats: "C", sharps: "C", camelot: "8B", openkey: "1d" },
  { flats: "Db", sharps: "C#", camelot: "3B", openkey: "8d" },
  { flats: "D", sharps: "D", camelot: "10B", openkey: "3d" },
  { flats: "Eb", sharps: "D#", camelot: "5B", openkey: "10d" },
  { flats: "E", sharps: "E", camelot: "12B", openkey: "5d" },
  { flats: "F", sharps: "F", camelot: "7B", openkey: "12d" },
  { flats: "Gb", sharps: "F#", camelot: "2B", openkey: "7d" },
  { flats: "G", sharps: "G", camelot: "9B", openkey: "2d" },
  { flats: "Ab", sharps: "G#", camelot: "4B", openkey: "9d" },
  { flats: "A", sharps: "A", camelot: "11B", openkey: "4d" },
  { flats: "Bb", sharps: "A#", camelot: "6B", openkey: "11d" },
  { flats: "B", sharps: "B", camelot: "1B", openkey: "6d" },
  { flats: "Am", sharps: "Am", camelot: "8A", openkey: "1m" },
  { flats: "Bbm", sharps: "A#m", camelot: "3A", openkey: "8m" },
  { flats: "Bm", sharps: "Bm", camelot: "10A", openkey: "3m" },
  { flats: "Cm", sharps: "Cm", camelot: "5A", openkey: "10m" },
  { flats: "Dbm", sharps: "C#m", camelot: "12A", openkey: "5m" },
  { flats: "Dm", sharps: "Dm", camelot: "7A", openkey: "12m" },
  { flats: "Ebm", sharps: "D#m", camelot: "2A", openkey: "7m" },
  { flats: "Em", sharps: "Em", camelot: "9A", openkey: "2m" },
  { flats: "Fm", sharps: "Fm", camelot: "4A", openkey: "9m" },
  { flats: "Gbm", sharps: "F#m", camelot: "11A", openkey: "4m" },
  { flats: "Gm", sharps: "Gm", camelot: "6A", openkey: "11m" },
  { flats: "Abm", sharps: "G#m", camelot: "1A", openkey: "6m" },
] as const satisfies readonly KeyRow[];

/** The 24 canonical (flats-notation) key values. */
export type CanonicalKey = (typeof KEY_TABLE)[number]["flats"];

export const CANONICAL_KEYS: readonly CanonicalKey[] = KEY_TABLE.map(
  (r) => r.flats,
);

const byFlats = new Map<string, KeyRow>(KEY_TABLE.map((r) => [r.flats, r]));

export function isCanonicalKey(value: string): value is CanonicalKey {
  return byFlats.has(value);
}

/**
 * Format a canonical (flats) key for display per the global Key Display As
 * option (§3.5). Unknown/free-text values pass through unchanged — set-row
 * keys are manually typed and may hold anything.
 */
export function formatKey(canonical: string, notation: KeyNotation): string {
  const row = byFlats.get(canonical);
  if (!row) return canonical;
  return row[notation];
}

/** Background the §9 contrast rule adjusts against: --bg-row. */
export const KEY_CONTRAST_BACKGROUND = "#1A1A1A";
export const KEY_CONTRAST_MIN = 4.5;

const adjustedCache = new Map<string, string>();

/**
 * Display color for a canonical key with the §9 contrast rule applied:
 * palette entries that fall short of 4.5:1 on --bg-row are lightened
 * (hue preserved) to the minimum lightness that passes. Canonical palette
 * (lib/palette.ts) is untouched; this is the derived variant.
 *
 * Returns null when the key is unknown — callers fall back to --text-primary.
 */
export function keyDisplayColor(canonical: string): string | null {
  if (!isCanonicalKey(canonical)) return null;
  let hex = adjustedCache.get(canonical);
  if (!hex) {
    hex = ensureContrast(
      KEY_COLORS[canonical].hex,
      KEY_CONTRAST_BACKGROUND,
      KEY_CONTRAST_MIN,
    );
    adjustedCache.set(canonical, hex);
  }
  return hex;
}

/**
 * Color for key text honoring the Colorful Keys option (§3.5/§6.6):
 * on → per-key derived color; off (or unknown key) → null, meaning render in
 * --text-primary like any other cell. Color follows the key, not the notation.
 */
export function keyColor(
  canonical: string,
  colorfulKeys: boolean,
): string | null {
  if (!colorfulKeys) return null;
  return keyDisplayColor(canonical);
}

File: backend/app/keys.py
"""Canonical 24-key x 4-notation table (planning/03-ui-design.md §6.6).

Internal representation is flats-canonical (`Gbm`, 24 values). The other three
notations are pure display renames of the same key. This is a TABLE, not a
formula (hard requirement) — transcribed verbatim from the spec.
"""
from __future__ import annotations

NOTATIONS = ("flats", "sharps", "camelot", "openkey")

# (flats, sharps, camelot, openkey) — 12 majors then 12 minors, spec row order.
KEY_TABLE: list[tuple[str, str, str, str]] = [
    # majors
    ("C", "C", "8B", "1d"),
    ("Db", "C#", "3B", "8d"),
    ("D", "D", "10B", "3d"),
    ("Eb", "D#", "5B", "10d"),
    ("E", "E", "12B", "5d"),
    ("F", "F", "7B", "12d"),
    ("Gb", "F#", "2B", "7d"),
    ("G", "G", "9B", "2d"),
    ("Ab", "G#", "4B", "9d"),
    ("A", "A", "11B", "4d"),
    ("Bb", "A#", "6B", "11d"),
    ("B", "B", "1B", "6d"),
    # minors
    ("Am", "Am", "8A", "1m"),
    ("Bbm", "A#m", "3A", "8m"),
    ("Bm", "Bm", "10A", "3m"),
    ("Cm", "Cm", "5A", "10m"),
    ("Dbm", "C#m", "12A", "5m"),
    ("Dm", "Dm", "7A", "12m"),
    ("Ebm", "D#m", "2A", "7m"),
    ("Em", "Em", "9A", "2m"),
    ("Fm", "Fm", "4A", "9m"),
    ("Gbm", "F#m", "11A", "4m"),
    ("Gm", "Gm", "6A", "11m"),
    ("Abm", "G#m", "1A", "6m"),
]

FLATS_KEYS: tuple[str, ...] = tuple(row[0] for row in KEY_TABLE)

# flats key -> {notation: display value}
_BY_FLATS: dict[str, dict[str, str]] = {
    row[0]: dict(zip(NOTATIONS, row)) for row in KEY_TABLE
}

# any notation's value -> flats key (case-sensitive on notation values)
_TO_FLATS: dict[str, str] = {}
for _row in KEY_TABLE:
    for _val in _row:
        _TO_FLATS.setdefault(_val, _row[0])


def is_canonical(key: str) -> bool:
    return key in _BY_FLATS


def to_notation(flats_key: str, notation: str) -> str | None:
    """Render a canonical flats key in one of the 4 notations. None if unknown."""
    if notation not in NOTATIONS:
        raise ValueError(f"unknown notation: {notation!r}")
    entry = _BY_FLATS.get(flats_key)
    return entry[notation] if entry else None


def from_any_notation(value: str) -> str | None:
    """Map a value in any of the 4 notations back to the canonical flats key."""
    return _TO_FLATS.get(value.strip())
